### PR TITLE
Instance search

### PR DESCRIPTION
**Pull request by Robert-Watts**
By submitting a PR for this repository you accept the contributor license agreement. 

### Description

Add instance wide search functionality. 

**This screen grab is currently just a proof of concept and may change**
image

### Testing

> Describe how this can be tested by reviewers. Be specific about anything not tested and reasons why. If this library has unit and/or integration testing, tests should be added for new functionality and existing tests should complete without errors.
>
> Please include any manual steps for testing end-to-end or functionality not covered by unit/integration tests.
>
> Also include details of the environment this PR was developed in (language/platform/browser version).

- [ ] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [ ] I have added documentation for new/changed functionality in this PR or in the documentation repo
- [x] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used, if not `main`

### CODE REVIEW

File: html/common/libs/Search/SearchClients.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+{
+    /**
+     * Load the clients for searching
+     *
+     * @return array An array of search objects
+     */
+    public function load(): array
+    {
+        // The user cannot access Clients
+        if (!$this->AUTH->instancePermissionCheck(36)) {
```
Comment (Robert-Watts) [inline html/common/libs/Search/SearchClients.php:14]: @Jbithell

File: html/common/libs/Search/SearchLocations.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+{
+    /**
+     * Load the clients for searching
+     *
+     * @return array An array of search objects
+     */
+    public function load(): array
+    {
+        // The user cannot access Locations
+        if (!$this->AUTH->instancePermissionCheck(87)) {
```
Comment (Robert-Watts) [inline html/common/libs/Search/SearchLocations.php:14]: @Jbithell

File: html/common/libs/Search/SearchProjects.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+{
+    /**
+     * Load the projects that are accessible
+     *
+     * @return array An array of search objects
+     */
+    public function load(): array
+    {
+        // The user cannot access projects
+        if (!$this->AUTH->instancePermissionCheck(20)) {
```
Comment (Robert-Watts) [inline html/common/libs/Search/SearchProjects.php:14]: @Jbithell

File: html/common/libs/Search/SearchCMS.php [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        $subQuery->where("b.cmsPages_id IS NULL");
+        $subQuery->get ("cmsPagesDrafts a", null, "a.cmsPagesDrafts_data, a.cmsPages_id");
+
+
+        //Get page, combined with the sub query so we have the page and the content in the same database call
+        $this->DBLIB->where("instances_id",$this->instance['instances_id']);
+        $this->DBLIB->where("cmsPages_deleted",0);
+        $this->DBLIB->where("cmsPages_archived",0);
+        $this->DBLIB->join($subQuery, "page.cmsPages_id = content.cmsPages_id");
+        if ($this->AUTH->data['instance']["instancePositions_id"]) $this->DBLIB->where("(cmsPages_visibleToGroups IS NULL OR (FIND_IN_SET(  ? , cmsPages_visibleToGroups) > 0))", Array($this->AUTH->data['instance']["instancePositions_id"])); //If the user doesn't have a position - they're bstudios staff
```
Comment (Robert-Watts) [inline html/common/libs/Search/SearchCMS.php:48]: @Jbithell

**Review by Jbithell — APPROVED** [review]

File: html/admin/search.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+$PAGEDATA['pageConfig'] = ["TITLE" => $PAGEDATA['term'] ? "Search result for: " . $PAGEDATA['term'] : "Search" , "BREADCRUMB" => false];
+
+
+/**
+ * Define a custom twig function to use a swtich case to generate the url the result should resolve to
+ */
+$function = new TwigFunction('generate_result_url', function ($result) {
+    switch ($result['type']){
+        case "project":
+            return "/project?id=".$result['data']['projects_id'];
```
Comment (Robert-Watts) [inline html/admin/search.php:21]: @Jbithell This just thrown an error in production

**Pull request merged**